terraform/node_output.go

```diff
@@ -83,19 +83,65 @@ func (n *NodeApplyableOutput) EvalTree() EvalNode {
 				},
 			},
 			&EvalOpFilter{
-				Ops: []walkOperation{walkRefresh, walkPlan, walkApply, walkValidate},
+				Ops: []walkOperation{walkRefresh, walkPlan, walkApply, walkValidate, walkDestroy, walkPlanDestroy},
 				Node: &EvalWriteOutput{
 					Name:      n.Config.Name,
 					Sensitive: n.Config.Sensitive,
 					Value:     n.Config.RawConfig,
 				},
 			},
-			&EvalOpFilter{
-				Ops: []walkOperation{walkDestroy, walkPlanDestroy},
-				Node: &EvalDeleteOutput{
-					Name: n.Config.Name,
-				},
-			},
 		},
 	}
 }
+
+// NodeDestroyableOutput represents an output that is "destroybale":
+// its application will remove the output from the state.
+type NodeDestroyableOutput struct {
```

Comment:
We have a special behavior for outputs under `-target` that ensures that they stay in the graph as long as any dependency is present. Do you think we need to do the same for destroy here too?

My initial instinct is yes, so that if I run `terraform destroy -target=aws_instance.foo` we'd still visit an output whose expression is `"${aws_instance.foo.id}"` and remove it, though it's _tricky_ if the expression is instead `"${aws_instance.foo.id} ${aws_instance.bar.id}"`.

In that multi-dependency case, I suppose maybe it's _right_ that we'd remove it from state anyway, because the expression is no longer resolvable anyway (there would be an error if we tried to eval it). Not totally sure, though... :thinking: 

BTW, the `TargetDownstream` function implemented by `NodeApplyableOutput` is what triggers this special behavior, and so that's what we'd add here if we want to retain the destroy nodes too.

Reply:
I _thought_ this was doing the right thing here, but now you have me questioning whether I understood the TargetDownstream code correctly. Let me run some tests to confirm here.

If we intend to remove all but the root module outputs from the state, it's kind of a moot point, since the targeted destroy is going to at least break the output temporarily until there's another apply to reify the missing resources for interpolation.

I think removing the root output is OK too. If a remote state is looking for it, it's probably better that it appear as missing and require fixing rather than using a known stale value.

Reply:
Descendants, Ancestors, Upstream, Downstream, (Up|Down|In|Out)Edges 🤕 

You're right, this was missed in some cases by targeting. `GraphNodeTargetDownstream` is an easy fix, since the dependency set of the destroy node is just the output and its ancestors.